src/bencode/decode.rs

```diff
@@ -1,51 +1,80 @@
+use std::collections::BTreeMap;
 use bencode::{Bencode, BString, BInt, BList, BDict, DecodeError, DecodeErrorKind};
 
-pub fn bstring_decode(bytes: Vec<u8>) -> Result<(BString, Vec<u8>), DecodeError> {
-    let mut length_str = String::new();
-    let mut position = 0u64;
-    if bytes.len() > 0 && (bytes[0] as char).is_digit(10) {
-        'topl: for ch in &bytes {
-            position += 1;
-            if (*ch as char) == ':' {
-                for byte in &bytes {
-                    if (*byte as char) != ':' {
-                        length_str.push((*byte as char));
-                    } else {
-                        break 'topl;
-                    }
-                }
-            }
-        }
+pub fn belement_decode(bytes: &[u8], position: &mut usize) -> Result<Bencode, DecodeError> {
+    if bytes.len() == 0 || *position >= bytes.len() {
+        return Err(DecodeError {
+            position: Some(*position as u64),
+            kind: DecodeErrorKind::EndOfStream,
+        });
+    }
+
+    if bytes[*position] == 'i' as u8 {
+        let result = try!(bint_decode(bytes, position));
+        Ok(Bencode::BInt(result))
+    } else if bytes[*position] == 'l' as u8 {
+        let result = try!(blist_decode(bytes, position));
+        Ok(Bencode::BList(result))
+    } else if bytes[*position] == 'd' as u8 {
+        let result = try!(bdict_decode(bytes, position));
+        Ok(Bencode::BDict(result))
     } else {
+        let result = try!(bstring_decode(bytes, position));
+        Ok(Bencode::BString(result))
+    }
+}
+
+pub fn bstring_decode(bytes: &[u8], position_arg: &mut usize) -> Result<BString, DecodeError> {
+    let mut position = *position_arg;
+    if !(bytes[position] as char).is_numeric() {
         return Err(DecodeError {
-            position: Some(position),
+            position: Some(position as u64),
             kind: DecodeErrorKind::InvalidString,
         });
     }
-    let length_int = try!(length_str.parse::<usize>());
-    let string_data = &bytes[length_str.len() + 1..(length_int + length_str.len()) + 1];
-    let remaining = bytes[string_data.len() + length_str.len() + 1..].to_vec();
-    let bstring = BString::new(string_data);
-    Ok((bstring, remaining))
+    while (bytes[position] as char).is_numeric() {
+        position += 1;
+    }
+    let len = &String::from_utf8(bytes[*position_arg..position]
+            .iter()
+            .map(|&x| x)
+            .collect::<Vec<u8>>())
+        .unwrap()
```

Comment:
The code looks a bit poopy, but a check is done on the bytes going into the string to make sure they are numeric 😂. I believe this means a panic from unwrap should never happen... but I like the idea of putting in error handling for it anyway. I think I want to improve this function even more: firstly by changing the is_numeric check to only allow ASCII ([is_numeric returns true for whole classes of unicode chars](https://doc.rust-lang.org/std/primitive.char.html#method.is_numeric). And secondly by verifying the bytes in the string to make sure they are UTF-8.